pennylane/templates/state_preparations.py

```diff
@@ -0,0 +1,315 @@
+        j = int(np.ceil(i * 2 ** (-k)))
+        s_condition = 2 ** (k - 1) * (2 * j - 1)
+        s_i = 1.0 if i > s_condition else -1.0
+        alpha_z_k[j - 1, 0] = alpha_z_k[j - 1, 0] + s_i * om / 2 ** (k - 1)
+
+    return alpha_z_k
+
+
+def _get_alpha_y(a, n, k):
+    """Computes the rotation angles alpha for the z-rotations
+
+    Args:
+        omega (float): phase of the input
+        n (int): total number of qubits
+        k (int): current qubit
+
+    Returns:
+        a sparse vector representing :math:`\alpha^y_k`
+    """
+    alpha = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
+
+    numerator = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
+    denominator = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
+
+    for (i, _), e in a.items():
+        j = int(math.ceil((i + 1) / 2 ** k))
+        l = (i + 1) - (2 * j - 1) * 2 ** (k - 1)
+        is_part_numerator = 1 <= l <= 2 ** (k - 1)
+
+        if is_part_numerator:
+            numerator[j - 1, 0] += e * e
+        denominator[j - 1, 0] += e * e
+
+    for (j, _), e in numerator.items():
+        numerator[j, 0] = np.sqrt(e)
+    for (j, _), e in denominator.items():
+        denominator[j, 0] = 1 / np.sqrt(e)
+
+    pre_alpha = numerator.multiply(denominator)  # type: sparse.csr_matrix
+    for (j, _), e in pre_alpha.todok().items():
+        alpha[j, 0] = 2 * np.arcsin(e)
+
+    return alpha
+
+
+def MottonenStatePreparation(state_vector, wires):
+    r"""
+    Prepares an arbitrary state on the given wires using a decomposition into gates developed
+    by Möttönen et al. :cite:`mottonen2005transformation`.
+
+    This code is adapted from code written by Carsten Blank for PennyLane-Qiskit.
+
+    Args:
+        state_vector (array): Input array of shape ``(2^N,)``, where N is the number of wires
+            the state preparation acts on. ``N`` must be smaller or equal to the total
+            number of wires.
+        wires (Sequence[int]): sequence of qubit indices that the template acts on
+    """
+
+    if not isinstance(wires, Iterable):
+        raise ValueError("Wires must be passed as a list of integers; got {}.".format(wires))
+
+    n = len(wires)
+
+    if not len(state_vector) == 2 ** n:
+        raise ValueError(
+            "Number of entries in the state vector must be equal to 2 to the power of the number of wires, which is {}; "
+            "got {}.".format(2 ** n, len(state_vector))
+        )
+
+    probability_sum = np.sum(np.abs(state_vector) ** 2)
+
+    if not np.isclose(probability_sum, 1.0, atol=1e-3):
+        raise ValueError(
+            "State vector probabilities have to sum up to 1.0, got {}".format(probability_sum)
+        )
+
+    # Change ordering of indices, original code was for IBM machines
+    reorder_indices = np.ravel_multi_index(
+        np.unravel_index(range(2 ** n), [2] * n, order="C"), [2] * n, order="F"
+    )
+
+    state_vector = np.array(state_vector)[reorder_indices, np.newaxis]
```

Comment:
A cleaner way of re-ordering the qubit order is to take the transpose:
```python
state_vector = state_vector.reshape([2] * n).T.flatten()
```

Reply:
I actually wanted to know how I can shorten this. Thanks!